### PR TITLE
Nullable

### PR DESCRIPTION
**Pull request by elharo**
some work on #112

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java [hunk 1]
```diff
@@ -50,6 +51,9 @@ private AppEngineDescriptor(Document document) {
    * @throws SAXException malformed XML
    */
   public static AppEngineDescriptor parse(InputStream in) throws IOException, SAXException {
+    if (in == null) {
+      throw new IOException("Null input");
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java]: Is IOException right?, given this is a programmer error, not an IO system error.

The nullaway plugin assumes without a `@Nullable` annotation, the parameter must not be null. So the person calling `parse` is responsible for the null check, or errorprone-nullaway will error out.

**Comment by elharo** [comment]
Failures are in test classes where we're deliberately triggering NullPointerExceptions. Trying to figure out how to exclude those for now.

**Comment by elharo** [comment]
Once again:

```

[WARN] JavaNCSS got an error while parsing the java file /home/travis/build/GoogleCloudPlatform/appengine-plugins-core/src/main/java/com/google/cloud/tools/managedcloudsdk/components/WindowsBundledPythonCopier.java

ParseException in STDIN
```

Not sure what's going on here.

**Comment by elharo** [comment]
Apparently a known issue:

https://github.com/cobertura/cobertura/issues/176

We might have to choose between cobertura and lambdas.

**Comment by briandealwis** [comment]
FWIW, jacoco doesn't seem to have issues with lambdas.

File: src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java [hunk 2]
```diff
@@ -50,6 +51,9 @@ private AppEngineDescriptor(Document document) {
    * @throws SAXException malformed XML
    */
   public static AppEngineDescriptor parse(InputStream in) throws IOException, SAXException {
+    if (in == null) {
```
Comment (briandealwis) [inline src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java]: Is this better than `Preconditions.checkNotNull()`?  Or annotating either the package or the class with `ParametersAreNonnullByDefault`?

File: src/main/java/com/google/cloud/tools/libraries/json/CloudLibrary.java [hunk 3]
```diff
@@ -22,17 +22,17 @@
...
+  @Nullable private String serviceName;
+  @Nullable private List<String> serviceRoles;
+  @Nullable private String documentation;
+  @Nullable private String description;
+  @Nullable private String icon;
+  @Nullable private List<String> transports;
+  @Nullable private List<CloudLibraryClient> clients;
 
-  /** Prevents instantiation. */
+  /** Prevents direct instantiation. GSON instantiates these objects using dark magic. */
```
Comment (briandealwis) [inline src/main/java/com/google/cloud/tools/libraries/json/CloudLibrary.java]: Nice :-)

File: src/main/java/com/google/cloud/tools/managedcloudsdk/command/CommandExitException.java [hunk 4]
```diff
@@ -16,28 +16,41 @@
...
-  private final String errLog;
+  @Nullable private String errorLog;
+
+  /**
+   * Create a new exception.
+   *
+   * @param exitCode the process exit code
+   * @param errorLog additional loggable error information, can be {@code null}
+   */
+  public CommandExitException(int exitCode, String errorLog) {
```
Comment (briandealwis) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/command/CommandExitException.java]: Doesn't `errorLog` need `@Nullable`?

File: pom.xml [hunk 5]
```diff
@@ -149,19 +149,34 @@
...
+               <artifactId>nullaway</artifactId>
+               <version>0.4.4</version>
+            </path>
+          </annotationProcessorPaths>
+          <compilerArgs>
+            <arg>-Xep:NullAway:ERROR</arg>    
+            <arg>-XepOpt:NullAway:AnnotatedPackages=com.google.cloud.tools.appengine,com.google.cloud.tools.libraries,com.google.cloud.tools.managedcloudsdk,com.google.cloud.tools.io,com.google.cloud.tools.project</arg>
+            <arg>-XepOpt:NullAway:UnannotatedSubPackages=com.google.cloud.tools.appengine.api.[a-zA-Z.]*,com.google.cloud.tools.appengine.cloudsdk.[a-zA-Z.]*</arg>
+<arg>-XepOpt:NullAway:ExcludedFieldAnnotations=org.mockito.Mock,org.junit.runners.Parameterized</arg>
+<arg>-XepOpt:NullAway:ExcludedClasses=com.google.cloud.tools.appengine.cloudsdk.CloudSdkAppEngineDevServer1Test,com.google.cloud.tools.appengine.cloudsdk.CloudSdkAppEngineDevServer2Test,com.google.cloud.tools.managedcloudsdk.install.InstallerFactoryTest,com.google.cloud.tools.appengine.AppEngineDescriptorTest</arg>
```
Comment (briandealwis) [inline pom.xml]: This seems to be very hard-to-maintain :-(

Comment (elharo) [reply pom.xml]: If it fails, then the build fails so early warning.

File: src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java (hunk 2 above)
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java]: ParametersAreNonnullByDefault is very different.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/command/CommandExitException.java (hunk 4 above)
Comment (elharo) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/command/CommandExitException.java]: No since this usage isn't nullable.

Comment (briandealwis) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/command/CommandExitException.java]: Correct the comment then?

File: pom.xml [hunk 6]
```diff
@@ -212,7 +226,7 @@
           <execution>
             <phase>validate</phase>
             <goals>
-              <goal>check</goal>
+              <goal>format</goal>
```
[the remark this answers is not in the document]
Comment (elharo) [reply pom.xml]: out for review

File: src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java [hunk 7]
```diff
@@ -114,12 +121,18 @@ public boolean isJava8() throws AppEngineException {
    *
    * @return a map representing the environment variable settings in the appengine-web.xml
    */
+  // todo this should probably return an empty map instead per Effective Java
+  @Nullable
   public Map<String, String> getEnvironment() throws AppEngineException {
     Node environmentParentNode = getNode(document, "appengine-web-app", "env-variables");
-    return getAttributeMap(environmentParentNode, "env-var", "name", "value");
+    if (environmentParentNode != null) {
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java]: done

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java [hunk 8]
```diff
@@ -241,21 +244,26 @@ public void validateAppEngineJavaComponents()
...
+    @Nullable private List<CloudSdkResolver> resolvers;
     private Path javaHomePath = Paths.get(System.getProperty("java.home"));
 
     /**
      * The home directory of Google Cloud SDK.
      *
      * @param sdkPath the root path for the Cloud SDK
      */
-    public Builder sdkPath(Path sdkPath) {
+    public Builder sdkPath(@Nullable Path sdkPath) {
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: OK, reverted

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAuth.java [hunk 9]
```diff
@@ -41,7 +41,7 @@ public CloudSdkAuth(GcloudRunner runner) {
   }
 
   /**
-   * Logs into the cloud sdk with a specific user (does not retrigger auth flow if user is already
+   * Logs into the cloud SDK with a specific user (does not retrigger auth flow if user is already
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkAuth.java]: done

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkOutOfDateException.java [hunk 10]
```diff
@@ -18,14 +18,15 @@
...
+import com.google.common.base.Preconditions;
 import javax.annotation.Nullable;
 
 /** The Cloud SDK that was found is too old (generally before 133.0.0). */
 public class CloudSdkOutOfDateException extends AppEngineException {
 
   private static final String MESSAGE = "Cloud SDK versions before %s are not supported";
 
-  private CloudSdkVersion installedVersion;
+  @Nullable private CloudSdkVersion installedVersion;
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkOutOfDateException.java]: It depends on which constructor you call.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/DevAppServerRunner.java [hunk 11]
```diff
@@ -100,7 +101,7 @@ void runV1(
       List<String> jvmArgs,
       List<String> args,
       Map<String, String> environment,
-      File workingDirectory)
+      @Nullable File workingDirectory)
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/cloudsdk/DevAppServerRunner.java]: done

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/GcloudRunner.java [hunk 12]
```diff
@@ -62,7 +63,7 @@
     this.processHandler = processHandler;
   }
 
-  void run(List<String> arguments, File workingDirectory)
+  void run(List<String> arguments, @Nullable File workingDirectory)
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/cloudsdk/GcloudRunner.java]: done

File: src/main/java/com/google/cloud/tools/managedcloudsdk/command/AsyncByteConsumer.java [hunk 13]
```diff
@@ -50,7 +51,7 @@
       ByteHandler byteHandler,
       ListeningExecutorService executorService,
       SettableFuture<String> result) {
-    this.byteHandler = byteHandler;
+    this.byteHandler = Preconditions.checkNotNull(byteHandler);
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/command/AsyncByteConsumer.java]: fixed

File: src/main/java/com/google/cloud/tools/managedcloudsdk/components/WindowsBundledPythonCopier.java [hunk 14]
```diff
@@ -55,12 +55,27 @@
...
 
     if (isUnderTempDirectory(pythonExePath, System.getenv())) {
-      Runtime.getRuntime().addShutdownHook(new Thread(() -> deleteCopiedPython(pythonExePath)));
+      Runtime.getRuntime().addShutdownHook(new Thread(new PythonDeleter(pythonExePath)));
     }
 
     return ImmutableMap.of("CLOUDSDK_PYTHON", pythonExePath);
   }
 
+  // WARNING: do not turn this into a lambda. JavaNCSS can't handle that.
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/components/WindowsBundledPythonCopier.java]: done

File: src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java [hunk 15]
```diff
@@ -117,18 +117,16 @@ public boolean isJava8() throws AppEngineException {
...
    */
-  // todo this should probably return an empty map instead per Effective Java
-  @Nullable
   public Map<String, String> getEnvironment() throws AppEngineException {
     Node environmentParentNode = getNode(document, "appengine-web-app", "env-variables");
     if (environmentParentNode != null) {
       return getAttributeMap(environmentParentNode, "env-var", "name", "value");
     }
-    return null;
+    return new HashMap<>();
```
Comment (briandealwis) [inline src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java]: Collections.emptyMap()

Comment (elharo) [reply src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java]: I don't think we should have a method that sometimes returns a mutable map and sometimes returns an immutable map depending on which code path is taken. furshlugginer Collections API :-(

**Pull request merged**